Add properties and methods to SR content items and templates

hackermd opened the pull request:
This will allow consistent access to values of content items independent of their value type. The only exception are content items of type CONTAINER, which I would argue don't have a value.

hackermd commented on src/highdicom/sr/value_types.py:
```diff
@@ -378,6 +440,11 @@ def __init__(
         self.UID = value
 
+    @property
+    def value(self) -> str:
```
Addressed with d3b1fdef30f67271082d525c50203d22ef4a5b71

hackermd commented:
@CPBridge I rebased this branch onto the HEAD of master (including the new import logic). I had to resolve a couple of merge conflicts and hope everything is still in order.

CPBridge commented on src/highdicom/sr/templates.py:
```diff
@@ -1366,6 +2131,133 @@ def __init__(
+        dataset = sequence[0]
+        if dataset.ValueType != ValueTypeValues.CONTAINER.value:
+            raise ValueError(
+                'Item #1 of sequence is not an appropropriate SR Content Item '
```
spellelling
```suggestion
                'Item #1 of sequence is not an appropriate SR Content Item '
```

CPBridge commented on src/highdicom/sr/templates.py:
```diff
@@ -1366,6 +2131,133 @@ def __init__(
+            )
+        if get_coded_name(dataset) != codes.DCM.MeasurementGroup:
+            raise ValueError(
+                'Item #1 of sequence is not an appropropriate SR Content Item '
```
```suggestion
                'Item #1 of sequence is not an appropriate SR Content Item '
```

CPBridge commented on src/highdicom/sr/templates.py:
```diff
@@ -1935,20 +2983,427 @@ def __init__(
+        dataset = sequence[0]
+        if dataset.ValueType != ValueTypeValues.CONTAINER.value:
+            raise ValueError(
+                'Item #1 of sequence is not an appropropriate SR Content Item '
```
```suggestion
                'Item #1 of sequence is not an appropriate SR Content Item '
```

CPBridge commented on src/highdicom/sr/templates.py:
```diff
@@ -1935,20 +2983,427 @@ def __init__(
+            )
+        if dataset.ContentTemplateSequence[0].TemplateIdentifier != '1500':
+            raise ValueError(
+                'Item #1 of sequence is not an appropropriate SR Content Item '
```
```suggestion
                'Item #1 of sequence is not an appropriate SR Content Item '
```

CPBridge commented on src/highdicom/sr/templates.py:
```diff
@@ -1935,20 +2983,427 @@ def __init__(
+
+        Returns
+        -------
+        List[highdicom.sr.templates.ObserverContext]
```
```suggestion
        List[highdicom.sr.ObserverContext]
```

CPBridge commented on src/highdicom/sr/templates.py:
```diff
@@ -1935,20 +2983,427 @@ def __init__(
+
+        Returns
+        -------
+        highdicom.sr.templates.MeasurementsReport
```
```suggestion
        highdicom.sr.MeasurementsReport
```

CPBridge commented on src/highdicom/sr/templates.py:
```diff
@@ -1935,20 +2983,427 @@ def __init__(
+
+    @classmethod
+    def from_sequence(cls, sequence: Sequence[Dataset]) -> 'MeasurementReport':
+        """Construct instance from a sequence of datasets.
```
"instance" is maybe a confusing word to use here?

hackermd replied:
I agree. That term could be confusing. What would you suggest? How about "Construct instance **of the class**."

CPBridge replied:
`Construct object` or `Construct MeasurementReport`?

hackermd replied:
Addressed via 505ea884ab767ff872222c35f207f564874bc17d

CPBridge commented on src/highdicom/sr/templates.py:
```diff
@@ -1935,20 +2983,427 @@ def __init__(
+
+        Returns
+        -------
+        List[highdicom.sr.templates.SubjectContext]
```
```suggestion
        List[highdicom.sr.SubjectContext]
```

CPBridge commented on src/highdicom/sr/templates.py:
```diff
@@ -1935,20 +2983,427 @@ def __init__(
+
+        Returns
+        -------
+        List[highdicom.sr.templates.PlanarROIMeasurementsAndQualitativeEvaluations]
```
```suggestion
        List[highdicom.sr.PlanarROIMeasurementsAndQualitativeEvaluations]
```

CPBridge commented on src/highdicom/sr/templates.py:
```diff
@@ -1935,20 +2983,427 @@ def __init__(
+            Union[GraphicTypeValues, GraphicTypeValues3D]
+        ] = None
+    ) -> List[PlanarROIMeasurementsAndQualitativeEvaluations]:
+        """Get imaging measurements of planar image regions of interest.
+
+        Finds (and optionally filters) content items contained in the
+        CONTAINER content item "Measurement group" as specified by TID 1410
+        "Planar ROI Measurements and Qualitative Evaluations".
```
I think this should make it clearer that this returns measurement groups rather than measurements themselves. (same for volumetric and image measurements)

hackermd replied:
After renaming the methods, do you think this would still need further clarification?

CPBridge replied:
I would still rephrase slightly to avoid confusion 
```suggestion
        """Get imaging measurement groups of planar image regions of interest.

        Finds (and optionally filters) content items contained in the
        CONTAINER content item "Measurement group" as specified by TID 1410
        "Planar ROI Measurements and Qualitative Evaluations".
```

CPBridge commented on src/highdicom/sr/templates.py:
```diff
@@ -1935,20 +2983,427 @@ def __init__(
+
+        Returns
+        -------
+        List[highdicom.sr.templates.VolumetricROIMeasurementsAndQualitativeEvaluations]
```
```suggestion
        List[highdicom.sr.VolumetricROIMeasurementsAndQualitativeEvaluations]
```

CPBridge commented on src/highdicom/sr/templates.py:
```diff
@@ -1935,20 +2983,427 @@ def __init__(
+
+        Returns
+        -------
+        List[highdicom.sr.templates.MeasurementsAndQualitativeEvaluations]
```
```suggestion
        List[highdicom.sr.MeasurementsAndQualitativeEvaluations]
```

CPBridge commented on src/highdicom/sr/templates.py:
```diff
@@ -1706,6 +2622,41 @@ def __init__(
+    def roi(
+        self
+    ) -> Union[VolumeSurface, List[ImageRegion3D], List[ImageRegion], None]:
+        """Union[highdicom.sr.content.VolumeSurface, List[highdicom.sr.content.ImageRegion], List[highdicom.sr.content.ImageRegion3D]], None]:
```
```suggestion
        """Union[highdicom.sr.VolumeSurface, List[highdicom.sr.ImageRegion], List[highdicom.sr.ImageRegion3D]], None]:
```

CPBridge commented on src/highdicom/sr/templates.py:
```diff
@@ -1618,6 +2510,30 @@ def __init__(
+
+    @property
+    def roi(self) -> Union[ImageRegion, ImageRegion3D, None]:
+        """Union[highdicom.sr.content.ImageRegion, highdicom.sr.content.ImageRegion3D], None]:
```
```suggestion
        """Union[highdicom.sr.ImageRegion, highdicom.sr.ImageRegion3D], None]:
```

CPBridge commented on src/highdicom/sr/templates.py:
```diff
@@ -1366,6 +2131,133 @@ def __init__(
+
+    @property
+    def qualitative_evaluations(self) -> List[CodeContentItem]:
+        """List[highdicom.sr.value_types.CodeContentItem]: qualitative
```
```suggestion
        """List[highdicom.sr.CodeContentItem]: qualitative
```

CPBridge commented on src/highdicom/sr/templates.py:
```diff
@@ -1366,6 +2131,133 @@ def __init__(
+        )
+
+    @property
+    def qualitative_evaluations(self) -> List[CodeContentItem]:
```
Return the new `QualitativeEvaluation` class?

hackermd replied:
Not sure. It would then be logical to have `measurements` return `List[Measurement]`. However, a `Measurement` is a `Template` (and hence a `ContentSequence`), while `QualitativeEvaluation` is a `CodeContentItem`. That may be confusing. Thoughts?

CPBridge replied:
Oh boy... you've made me go and look more closely at the way measurement parsing has been implemented and opened up a can of worms.

I think my opinion here is that we should do what we can to mask the different between sequences and items from all but the most advanced users. It is only an important distinction to me when you are actually trying to piece the building blocks together the right way, which users should never be doing themselves. Consequently I think my suggestion would be that the `measurements` property or `get_measurements` method (whichever we go with) should probably return `Measurement` and furthermore `Measurement` should be given various properties to access its name, value, unit, tracking id, etc etc... Then the user really doesn't need to care whether what they get back is a sequence or an item.

I think the current situation in which when constructing a measurement the user provides, or may provide, all this extra information, but then on the parsing side only gets back the `NumContentItem` is quite confusing. There's currently no wway to access the `unit` at all, right? But that could be very important in many situations. Not much use getting the measurement without the unit, in my opinion.

I have done a similar thing to other sequences elsewhere, see `AlgorithmIdentifcationSequence` [here](https://github.com/MGHComputationalPathology/highdicom/blob/94d7c9d85ce5cf46b5230170ea8abe08782b0ff6/src/highdicom/content.py#L27),although that case is somewhat simpler.

To bring this back to the original point, I would also suggest returning `QualitativeEvaluation` here. I think generally having symmetry with encoding and decoding is going to make it easier for people to understand.

Sorry this complicates things when it felt like we were nearly there..., but I think this is a really important part to get right. What do you think?

hackermd replied:
> I think my opinion here is that we should do what we can to mask the different between sequences and items from all but the most advanced users. It is only an important distinction to me when you are actually trying to piece the building blocks together the right way, which users should never be doing themselves. Consequently I think my suggestion would be that the measurements property or get_measurements method (whichever we go with) should probably return Measurement and furthermore Measurement should be given various properties to access its name, value, unit, tracking id, etc etc... Then the user really doesn't need to care whether what they get back is a sequence or an item.

Or we make an exception and implement the measurement content item in `highdicom.sr.content` derived from `NumContentItem` rather than implementing the TID 300 "Measurment" template in `highdicom.sr.templates`?

That would be more intuitive and also circumvent some problems related the inclusion of TID 300 in other templates. TID 300 is included in TID 1501 "Measurement and Qualitative Evaluation Group", but neither in TID 1410 "Planar ROI Measurements and Qualitative Evaluations" nor in TID 1411 "Volumetric ROI Measurements and Qualitative Evaluations".

This would break backwards compatibility, but I think it would be worth it. What do you think?

CPBridge replied:
I don't think I understand your proposal. How could you implement [TID 300](http://dicom.nema.org/medical/dicom/current/output/chtml/part16/chapter_A.html#sect_TID_300), with units, qualitifers, tracking IDs, etc etc, as a single NumContentItem?

hackermd replied:
As discussed, I moved `QualitativeEvaluation` from `highdicom.sr.content` to `highdicom.sr.templates` and changed the base type from `CodeContentItem` to `Template` for consistency with `Measurement` (see 0b5ae3de258abb2c7c17ed3433fe607c9ece0d25). Now both `Measurement` and `QualitativeEvaluation` have a similar (awkward) interface.

CPBridge commented on src/highdicom/sr/templates.py:
```diff
@@ -1366,6 +2131,133 @@ def __init__(
+
+    @property
+    def measurements(self) -> List[NumContentItem]:
+        """List[highdicom.sr.value_types.NumContentItem]: measurements"""
```
```suggestion
        """List[highdicom.sr.NumContentItem]: measurements"""
```

CPBridge commented on src/highdicom/sr/templates.py:
```diff
@@ -1366,6 +2131,133 @@ def __init__(
+
+    @property
+    def finding_sites(self) -> List[FindingSite]:
+        """List[highdicom.sr.content.FindingSite]: finding sites"""
```
```suggestion
        """List[highdicom.sr.FindingSite]: finding sites"""
```

CPBridge commented on src/highdicom/sr/templates.py:
```diff
@@ -1366,6 +2131,133 @@ def __init__(
+
+    @property
+    def finding_type(self) -> Union[CodedConcept, None]:
+        """Union[highdicom.sr.coding.CodedConcept, None]: finding"""
```
```suggestion
        """Union[highdicom.sr.CodedConcept, None]: finding"""
```

CPBridge commented on src/highdicom/sr/templates.py:
```diff
@@ -1366,6 +2131,133 @@ def __init__(
+
+    @property
+    def method(self) -> Union[CodedConcept, None]:
+        """Union[highdicom.sr.coding.CodedConcept, None]: measurement method"""
```
```suggestion
        """Union[highdicom.sr.CodedConcept, None]: measurement method"""
```

CPBridge commented on src/highdicom/sr/templates.py:
```diff
@@ -1366,6 +2131,133 @@ def __init__(
+
+        Returns
+        -------
+        highdicom.sr.templates.MeasurementsAndQualitativeEvaluations
```
```suggestion
        highdicom.sr.MeasurementsAndQualitativeEvaluations
```

CPBridge commented on src/highdicom/sr/templates.py:
```diff
@@ -973,8 +1695,50 @@ def __init__(
+        SubjectContextSpecimen,
+        SubjectContextDevice
+    ]:
+        """Union[highdicom.sr.templates.SubjectContextFetus, highdicom.sr.templates.SubjectContextSpecimen, highdicom.sr.templates.SubjectContextDevice]:
```
```suggestion
        """Union[highdicom.sr.SubjectContextFetus, highdicom.sr.SubjectContextSpecimen, highdicom.sr.SubjectContextDevice]:
```

CPBridge commented on src/highdicom/sr/templates.py:
```diff
@@ -933,6 +1561,102 @@ def __init__(
+
+        Returns
+        -------
+        highdicom.sr.templates.SubjectContextDevice
```
```suggestion
        highdicom.sr.SubjectContextDevice
```

CPBridge commented on src/highdicom/sr/templates.py:
```diff
@@ -833,17 +1401,77 @@ def __init__(
+
+        Returns
+        -------
+        highdicom.sr.templates.SubjectContextSpecimen
```
```suggestion
        highdicom.sr.SubjectContextSpecimen
```

CPBridge commented on src/highdicom/sr/templates.py:
```diff
@@ -769,6 +1291,41 @@ def __init__(self, subject_id: str):
+
+        Returns
+        -------
+        highdicom.sr.templates.SubjectContextFetus
```
```suggestion
        highdicom.sr.SubjectContextFetus
```

CPBridge commented on src/highdicom/sr/templates.py:
```diff
@@ -742,6 +1242,28 @@ def __init__(
+        PersonObserverIdentifyingAttributes,
+        DeviceObserverIdentifyingAttributes,
+    ]:
+        """Union[highdicom.sr.templates.PersonObserverIdentifyingAttributes, highdicom.sr.templates.DeviceObserverIdentifyingAttributes]:
```
```suggestion
        """Union[highdicom.sr.PersonObserverIdentifyingAttributes, highdicom.sr.DeviceObserverIdentifyingAttributes]:
```

CPBridge commented on src/highdicom/sr/templates.py:
```diff
@@ -584,6 +891,92 @@ def __init__(
+
+        Returns
+        -------
+        highdicom.sr.templates.PersonObserverIdentifyingAttributes
```
```suggestion
        highdicom.sr.PersonObserverIdentifyingAttributes
```

CPBridge commented on src/highdicom/sr/value_types.py:
```diff
@@ -64,23 +139,67 @@ def __setattr__(self, name: str, value: Any) -> None:
+
+        Returns
+        -------
+        highdicom.sr.value_types.ContentItem
```
```suggestion
        highdicom.sr.ContentItem
```

CPBridge commented on src/highdicom/sr/value_types.py:
```diff
@@ -169,6 +288,45 @@ def insert(self, position: int, item: ContentItem) -> None:
+
+        Returns
+        -------
+        highdicom.sr.value_types.ContentSequence
```
```suggestion
        highdicom.sr.ContentSequence
```

CPBridge commented on src/highdicom/sr/value_types.py:
```diff
@@ -204,6 +362,34 @@ def __init__(
 
+    @property
+    def value(self) -> CodedConcept:
+        """highdicom.sr.coding.CodedConcept: coded concept"""
```
```suggestion
        """highdicom.sr.CodedConcept: coded concept"""
```

CPBridge commented on src/highdicom/sr/value_types.py:
```diff
@@ -204,6 +362,34 @@ def __init__(
+
+        Returns
+        -------
+        highdicom.sr.value_types.CodeContentItem
```
```suggestion
        highdicom.sr.CodeContentItem
```

CPBridge commented on src/highdicom/sr/value_types.py:
```diff
@@ -233,6 +419,29 @@ def __init__(
+
+        Returns
+        -------
+        highdicom.sr.value_types.PnameContentItem
```
```suggestion
        highdicom.sr.PnameContentItem
```

CPBridge commented on src/highdicom/sr/value_types.py:
```diff
@@ -262,6 +471,29 @@ def __init__(
+
+        Returns
+        -------
+        highdicom.sr.value_types.TextContentItem
```
```suggestion
        highdicom.sr.TextContentItem
```

CPBridge commented on src/highdicom/sr/value_types.py:
```diff
@@ -291,6 +523,41 @@ def __init__(
+
+        Returns
+        -------
+        highdicom.sr.value_types.TimeContentItem
```
```suggestion
        highdicom.sr.TimeContentItem
```

CPBridge commented on src/highdicom/sr/value_types.py:
```diff
@@ -320,6 +587,30 @@ def __init__(
+
+        Returns
+        -------
+        highdicom.sr.value_types.DateContentItem
```
```suggestion
        highdicom.sr.DateContentItem
```

CPBridge commented on src/highdicom/sr/value_types.py:
```diff
@@ -349,6 +640,48 @@ def __init__(
+
+        Returns
+        -------
+        highdicom.sr.value_types.DateTimeContentItem
```
```suggestion
        highdicom.sr.DateTimeContentItem
```

CPBridge commented on src/highdicom/sr/value_types.py:
```diff
@@ -378,6 +711,29 @@ def __init__(
+
+        Returns
+        -------
+        highdicom.sr.value_types.UIDRefContentItem
```
```suggestion
        highdicom.sr.UIDRefContentItem
```

CPBridge commented on src/highdicom/sr/value_types.py:
```diff
@@ -452,6 +808,39 @@ def __init__(
+
+        Returns
+        -------
+        highdicom.sr.value_types.NumContentItem
```
```suggestion
        highdicom.sr.NumContentItem
```

CPBridge commented on src/highdicom/sr/value_types.py:
```diff
@@ -493,6 +882,33 @@ def __init__(
+
+        Returns
+        -------
+        highdicom.sr.value_types.ContainerContentItem
```
```suggestion
        highdicom.sr.ContainerContentItem
```

CPBridge commented on src/highdicom/sr/value_types.py:
```diff
@@ -528,6 +944,30 @@ def __init__(
+
+        Returns
+        -------
+        highdicom.sr.value_types.CompositeContentItem
```
```suggestion
        highdicom.sr.CompositeContentItem
```

CPBridge commented on src/highdicom/sr/value_types.py:
```diff
@@ -579,6 +1019,30 @@ def __init__(
+
+        Returns
+        -------
+        highdicom.sr.value_types.ImageContentItem
```
```suggestion
        highdicom.sr.ImageContentItem
```

CPBridge commented on src/highdicom/sr/value_types.py:
```diff
@@ -672,10 +1133,44 @@ def __init__(
+
+        Returns
+        -------
+        highdicom.sr.value_types.ScoordContentItem
```
```suggestion
        highdicom.sr.ScoordContentItem
```

CPBridge commented on src/highdicom/sr/value_types.py:
```diff
@@ -758,6 +1253,36 @@ def __init__(
+
+        Returns
+        -------
+        highdicom.sr.value_types.Scoord3DContentItem
```
```suggestion
        highdicom.sr.Scoord3DContentItem
```

CPBridge commented on src/highdicom/sr/value_types.py:
```diff
@@ -817,3 +1342,33 @@ def __init__(
+
+        Returns
+        -------
+        highdicom.sr.value_types.TcoordContentItem
```
```suggestion
        highdicom.sr.TcoordContentItem
```

CPBridge commented on src/highdicom/sr/templates.py:
```diff
@@ -1935,20 +2983,427 @@ def __init__(
+        ----------
+        tracking_uid: str, optional
+            Unique tracking identifier
+        finding_type: Union[highdicom.sr.coding.CodedConcept, pydicom.sr.coding.Code], optional
+            Finding
+        finding_site: Union[highdicom.sr.coding.CodedConcept, pydicom.sr.coding.Code], optional
+            Finding site
+        graphic_type: Union[highdicom.sr.enum.GraphicTypeValues, highdicom.sr.enum.GraphicTypeValues3D], optional
```
```suggestion
        finding_type: Union[highdicom.sr.CodedConcept, pydicom.sr.coding.Code], optional
            Finding
        finding_site: Union[highdicom.sr.CodedConcept, pydicom.sr.coding.Code], optional
            Finding site
        graphic_type: Union[highdicom.sr.GraphicTypeValues, highdicom.sr.GraphicTypeValues3D], optional
```

CPBridge commented on src/highdicom/sr/templates.py:
```diff
@@ -1935,20 +2983,427 @@ def __init__(
+        ----------
+        tracking_uid: str, optional
+            Unique tracking identifier
+        finding_type: Union[highdicom.sr.coding.CodedConcept, pydicom.sr.coding.Code], optional
+            Finding
+        finding_site: Union[highdicom.sr.coding.CodedConcept, pydicom.sr.coding.Code], optional
+            Finding site
+        graphic_type: highdicom.sr.enum.GraphicTypeValues3D, optional
```
```suggestion
        finding_type: Union[highdicom.sr.CodedConcept, pydicom.sr.coding.Code], optional
            Finding
        finding_site: Union[highdicom.sr.CodedConcept, pydicom.sr.coding.Code], optional
            Finding site
        graphic_type: highdicom.sr.GraphicTypeValues3D, optional
```

CPBridge commented on src/highdicom/sr/templates.py:
```diff
@@ -1935,20 +2983,427 @@ def __init__(
+        ----------
+        tracking_uid: str, optional
+            Unique tracking identifier
+        finding_type: Union[highdicom.sr.coding.CodedConcept, pydicom.sr.coding.Code], optional
+            Finding
+        finding_site: Union[highdicom.sr.coding.CodedConcept, pydicom.sr.coding.Code], optional
```
```suggestion
        finding_type: Union[highdicom.sr.CodedConcept, pydicom.sr.coding.Code], optional
            Finding
        finding_site: Union[highdicom.sr.CodedConcept, pydicom.sr.coding.Code], optional
```

CPBridge reviewed (changes requested):
@hackermd very nice! This is going to be a huge quality of life improvement. I made a few suggestions on the code, but there are a couple of high level comments:

1. Naming of the `get_{planar_roi,volumetric_roi,image}_measurements` methods. These names suggest that they return the measurements themselves, which they don't. I think this could cause a lot of confusion. For example to actually access the measurements you need to do `get_planar_roi_measusurements[0].measurements`, which seems counter-intuitive. I suggest we rename these `get_planar_roi_measurement_groups`, etc, or maybe some abbreviated version of this.
2. The current implementation supports the use case of ROIs stored as graphic data quite nicely, but does not really consider the other cases, for example measurements and evaluations relating to a referenced segment. Should there be properties/methods on a measurement group that allow a user to conveniently find out whether a given measurement group relates to a graphic data ROI, a segment, a volume surface etc. Should the `get_x_roi_measurement` groups allow you to filter by the "type" of ROI (graphic data, segment, volume surface), and potentially filter for measurement groups that relate to a specific segment, for example. I think we should also add properties to access, for example, the referenced segment of a measurement group, or at least leave this possibility open for the future. Currently the `roi` property only returns graphic data ROIs, if any. This seems potentially problematic as I don't see how we would add functionality to get referenced segments (which are also ROIs, kinda) in the future if this name is taken already. I'm not sure what the right solution is just at the moment, but I'm happy to brainstorm. Maybe we don't solve all these issues now, but have a plan as to how this can evolve.
3. I'm not so sure about the idea that `.content` returns a copy of the content tree converted into highdicom types, i.e. lazy type conversion. I don't really see why this is preferable to simply converting everything inside the content tree to the relevant highdicom types in the top-level `from_dataset`, and have `.content` return the pre-calculated. There are primarily two reasons why I think the latter approach is better:
   * I think it is very likely that a user will call `.content` multiple times for any given loaded document and having to re-do this conversion every time seems like an unnecessary overhead. Having to perform this conversion every time therefore creates a big performance hit. At the very least I think we need to cache this after the first call (as I believe pydicom does with `.pixel_array`)
   * Conceptually, the `from_dataset` method should return an object that is indistinguishable from one that was created by the relevant constructor. For users, this means that in the two cases if the go poking around in the pydicom API in the two cases (object created from scratch, object created with `from_dataset`) everything looks the same. More importantly for us as developers adding future functionality to these classes, we don't want to have to worry about how the object was created and consequently whether the items in the content tree are highdicom types or plain pydicom Datasets/Sequences.
4. Performance. Currently this implementation is pretty slow when the number of ROIs becomes high. I tested with an example I created with `highdicom_experiments` that contains 1029 ROIs from the  lung nodule detection module. This is certainly larger than most SRs are likely to be, but perfectly plausible, especially for SM images etc, and is only a 3MB file. The `.content` property takes 10s to evaluate, and then the `get_planar_roi_measurements` takes a further 20, giving 30 seconds to run a simple query. Eagerly performing type conversion as above, or caching the `.content` result will help but is not going to completely solve the problem. I haven't looked into this in detail but I'd guess that all the `find_content_items` are slowing things down.
5. There are many places where the docstrings need to be changed to adopt the new "flatter" import hierarchy. This is needed to get the sphinx docs to hyperlink properly. I have corrected many of these but almost certainly missed some.

CPBridge commented on src/highdicom/sr/value_types.py:
```diff
@@ -64,23 +139,67 @@ def __setattr__(self, name: str, value: Any) -> None:
-        return self.ConceptNameCodeSequence[0]
+        """highdicom.sr.CodedConcept: coded name of the content item"""
+        ds = self.ConceptNameCodeSequence[0]
+        return CodedConcept(
+            value=ds.CodeValue,
+            scheme_designator=ds.CodingSchemeDesignator,
+            meaning=ds.CodeMeaning
+        )
```
I just realised that I already created this for the segmentation parsing functionality on that branch...

I suggest we may want to try merging both these into a dev branch before master to work out potential conflicts and make everything consistent between them before merging to master.

hackermd replied:
I added the `CodedConcept.from_dataset()` method to this branch as well. Not sure whether we need a separate dev branch. We could also just resolve the two PRs sequentially when merging them into master or merge one of the two branches into the other. Any preference?

hackermd replied:
> This pattern recurs a lot. Should we create a CodedConcept.from_dataset() method?

Addressed via cad8120da70ffee7016dfe9ad02312827ddd38bc

hackermd commented:
@CPBridge Thanks for your review and thoughtful feedback.

> Naming of the get_{planar_roi,volumetric_roi,image}_measurements methods. These names suggest that they return the measurements themselves, which they don't. I think this could cause a lot of confusion. For example to actually access the measurements you need to do get_planar_roi_measusurements[0].measurements, which seems counter-intuitive. I suggest we rename these get_planar_roi_measurement_groups, etc, or maybe some abbreviated version of this

Good point. I have been using the names of the templates in the standard. It's unfortunate that they are not all consistently include "group" in their names. I have followed your suggestion and renamed the methods to `*_groups` via b607a3b3dc86afcca79fabe9f5aaa611a3b925e9.


> The current implementation supports the use case of ROIs stored as graphic data quite nicely, but does not really consider the other cases, for example measurements and evaluations relating to a referenced segment. Should there be properties/methods on a measurement group that allow a user to conveniently find out whether a given measurement group relates to a graphic data ROI, a segment, a volume surface etc. Should the get_x_roi_measurement groups allow you to filter by the "type" of ROI (graphic data, segment, volume surface), and potentially filter for measurement groups that relate to a specific segment, for example. I think we should also add properties to access, for example, the referenced segment of a measurement group, or at least leave this possibility open for the future. Currently the roi property only returns graphic data ROIs, if any. This seems potentially problematic as I don't see how we would add functionality to get referenced segments (which are also ROIs, kinda) in the future if this name is taken already. I'm not sure what the right solution is just at the moment, but I'm happy to brainstorm. Maybe we don't solve all these issues now, but have a plan as to how this can evolve.

I have been having a hard time with this. The fact that the ROIs can be encoded "per value" as well as "per reference" complicates things quite a bit. I like the idea of allowing the caller to filter groups by ROI value type (SCOORD, SCOORD3D, IMAGE, COMPOSITE). Maybe we should change `graphic_type: ` to `roi_value_type`:

```python
    def get_volumetric_roi_measurement_groups(
        self,
        tracking_uid: Optional[str] = None,
        finding_type: Optional[Union[CodedConcept, Code]] = None,
        finding_site: Optional[Union[CodedConcept, Code]] = None,
        roi_value_type: Optional[ValueTypeValues] = None
    ) -> List[VolumetricROIMeasurementsAndQualitativeEvaluations]:
    ...
```

Alternatively, we could use the concept name instead of the value type, but I have the feeling that this could be more messy. What do you think?

> I'm not so sure about the idea that .content returns a copy of the content tree converted into highdicom types, i.e. lazy type conversion. I don't really see why this is preferable to simply converting everything inside the content tree to the relevant highdicom types in the top-level from_dataset, and have .content return the pre-converted root content item.

Agreed. Fixed that via da9f066193bb1ea524c1de9b910b2383c5033bc1. The current implementation is not pretty, but it's tricky, because the content object (e.g., `MeasurementReport`) does not actually exist on the `SOPClass` object.

> Performance. Currently this implementation is pretty slow when the number of ROIs becomes high

Yeah, I haven't implemented this in the most performance-optimized way and there is a lot of room for improvement. I wanted to get the API sorted out first. I think we will be able to improve performance quite a bit with caching information in lookup tables on the objects rather than rerunning nested loops multiple times. Will need look into it.

>  There are many places where the docstrings need to be changed to adopt the new "flatter" import hierarchy. This is needed to get the sphinx docs to hyperlink properly. I have corrected many of these but almost certainly missed some.

Good point. I will go through the docstrings and update the types accordingly. Anyways wanted to change the type annotations.

hackermd commented on src/highdicom/sr/templates.py:
```diff
@@ -1935,20 +2983,427 @@ def __init__(
+
+        return sequences
+
+    def get_image_measurments(
```
This has changed in the meantime. Addressed via d59788b5cc2cecabdd46e79c84f0f435b6c8ae79 instead.

hackermd commented on src/highdicom/sr/content.py:
```diff
@@ -730,6 +733,28 @@ def __init__(
+            name=codes.SCT.TopographicalModifier,
+            value_type=ValueTypeValues.CODE
+        )
+        if len(matches) > 0:
+            return matches[0].value
```
Throw an error or log a warning? There should only be one item according to the standard. If there is more than one, it may still be valuable to return the first item without stopping program execution. What do you think?

CPBridge replied:
No strong feelings, a warning is probably sufficient

hackermd replied:
Added log statements via 4475f05450ed030f1619dc741bc701e2f68b555b

hackermd commented on src/highdicom/sr/templates.py:
```diff
@@ -36,6 +43,306 @@
+    )
+    regions_in_space_items = find_content_items(
+        group_item,
+        name=codes.DCM.RegionInSpace,
```
I think this code has only recently been added to the standard and will require an update of the codedict in pydicom. For the time being, I have replaced it with a constant (see 70a79129617029f62d26e4630a8de38d27b0a892)

hackermd commented on src/highdicom/sr/templates.py:
```diff
@@ -584,6 +892,92 @@ def __init__(
+            for param, name in attr_codes:
+                if content_item.name == name:
+                    kwargs[param] = content_item.value
+        return PersonObserverIdentifyingAttributes(**kwargs)
```
I have accepted all of your suggestions, but just realized that this may actually a really bad idea. `cls` may not be the class for which the `classmethod` has been defined, but a derived class. Here, we really want to construct an object of the given type not an instance of a derived type. I think we should undo all of these changes

hackermd replied:
Realized that you were right about this.

hackermd commented on src/highdicom/sr/value_types.py:
```diff
@@ -452,6 +808,39 @@ def __init__(
+        except AttributeError:
+            return item.NumericValue
+
+    @property
+    def unit(self) -> CodedConcept:
+        """highdicom.sr.coding.CodedConcept: unit"""
+        item = self.MeasuredValueSequence[0]
+        return item.MeasurementUnitsCodeSequence[0]
```
> Can unit be None? The constructor's signature suggests yes but crashes if you actually try to pass None. This seems like a bug.

I think you are right. `value` and `unit` appear to be required:

http://dicom.nema.org/medical/dicom/current/output/chtml/part03/sect_10.22.html
http://dicom.nema.org/medical/dicom/current/output/chtml/part03/sect_C.18.html#sect_C.18.1

Not sure why I made them optional. Changed that via f34cfc182066259b898910bf2cfab7670718fb2f

hackermd replied:
>

I addressed that via 8c96ec996c21326b7fb10a94f8aa243fcc87757f (copied the `CodedConcept.from_dataset` implementation from https://github.com/MGHComputationalPathology/highdicom/blob/feature/seg_decoding/src/highdicom/sr/coding.py)

hackermd commented on src/highdicom/sr/value_types.py:
```diff
@@ -452,6 +808,39 @@ def __init__(
+        """highdicom.sr.coding.CodedConcept: unit"""
+        item = self.MeasuredValueSequence[0]
+        return item.MeasurementUnitsCodeSequence[0]
+
```
Added via 22a54b7fad90344f45e2ae0b09025052bc785a4e

hackermd commented on src/highdicom/sr/value_types.py:
```diff
@@ -528,6 +944,30 @@ def __init__(
+    def value(self) -> Tuple[str, str]:
+        """Tuple[str, str]: referenced SOP Class UID and SOP Instance UID"""
+        item = self.ReferencedSOPSequence[0]
+        return (item.ReferencedSOPClassUID, item.ReferencedSOPInstanceUID)
```
Fixed via 005fc3d8fbdd345cb8d1a788cd2ac7af7e867996

hackermd commented on src/highdicom/sr/value_types.py:
```diff
@@ -579,6 +1019,30 @@ def __init__(
+    def value(self) -> Tuple[str, str]:
+        """Tuple[str, str]: referenced SOP Class UID and SOP Instance UID"""
+        item = self.ReferencedSOPSequence[0]
+        return (item.ReferencedSOPClassUID, item.ReferencedSOPInstanceUID)
```
Fixed via 005fc3d8fbdd345cb8d1a788cd2ac7af7e867996

hackermd commented on src/highdicom/sr/value_types.py:
```diff
@@ -676,6 +1140,36 @@ def __init__(
 
+    @property
+    def value(self) -> np.ndarray:
+        """numpy.ndarray: spatial coordinates"""
```
Addressed via e8d50a41f1da201f89b7108ce1261be15e4a6715

hackermd commented on src/highdicom/sr/value_types.py:
```diff
@@ -676,6 +1140,36 @@ def __init__(
+        """numpy.ndarray: spatial coordinates"""
+        graphic_data = np.array(self.GraphicData)
+        n_points = len(graphic_data) / 2
+        return np.array(np.array_split(graphic_data, n_points))
```
Addressed via b5b3f7e203ebf419cf7adef94aa085662200de0d

hackermd commented on src/highdicom/sr/value_types.py:
```diff
@@ -758,6 +1252,36 @@ def __init__(
+        """numpy.ndarray: spatial coordinates"""
+        graphic_data = np.array(self.GraphicData)
+        n_points = len(graphic_data) / 3
+        return np.array(np.array_split(graphic_data, n_points))
```
Addressed via b5b3f7e203ebf419cf7adef94aa085662200de0d

hackermd commented on src/highdicom/sr/value_types.py:
```diff
@@ -169,6 +288,45 @@ def insert(self, position: int, item: ContentItem) -> None:
+                raise AttributeError(
+                    'Dataset is not an SR Content Item because it lacks '
+                    f'required attribute "RelationshipType":\n{dataset}'
+                )
```
Addressed via bb9fd8e85d5fc72bfeefecee0d5a9c14b6b83a0e

hackermd commented on src/highdicom/sr/value_types.py:
```diff
@@ -452,6 +808,39 @@ def __init__(
+
+        """
+        _assert_value_type(dataset, ValueTypeValues.NUM)
+        return super(NumContentItem, cls)._from_dataset(dataset)
```
This should be fixed with 8c96ec996c21326b7fb10a94f8aa243fcc87757f

hackermd commented on src/highdicom/sr/templates.py:
```diff
@@ -679,6 +1084,102 @@ def __init__(
             self.append(role_in_procedure_item)
 
+    @property
+    def uid(self) -> str:
```
Addressed via 979e3aee98307d160e36885087c121f134a2e35b

hackermd commented on src/highdicom/sr/templates.py:
```diff
@@ -1366,6 +2132,133 @@ def __init__(
+        return None
+
+    @property
+    def tracking_uid(self) -> Union[str, None]:
```
Addressed via 979e3aee98307d160e36885087c121f134a2e35b

hackermd commented on src/highdicom/sr/templates.py:
```diff
@@ -1366,6 +2132,133 @@ def __init__(
+        )
+        if len(matches) > 0:
+            return [FindingSite.from_dataset(m) for m in matches]
+        return None
```
Addressed via 7f9f7f46e3e49595518553a7a6622ae21a9c4688

hackermd commented on src/highdicom/sr/templates.py:
```diff
@@ -1836,7 +2788,8 @@ def __init__(
             name=title,
-            template_id='1500'
+            template_id='1500',
+            relationship_type=RelationshipTypeValues.CONTAINS
```
Yes! Has been fixed via e1f7caa8

hackermd commented on src/highdicom/sr/templates.py:
```diff
@@ -1935,20 +2888,515 @@ def __init__(
+        """  # noqa
+        root_item = self[0]
+        matches = [
+            (i + 1, item) for i, item in enumerate(root_item.ContentSequence)
```
Are you sure your suggestion is correct?

CPBridge replied:
nope...
```suggestion
            (i, item) for i, item in enumerate(root_item.ContentSequence, 1)
```

hackermd commented on src/highdicom/sr/templates.py:
```diff
@@ -1935,20 +2888,515 @@ def __init__(
+            if len(matches) == 0:
+                sequences.append(seq)
+            else:
+                if any(matches):
```
The performance definitely needs to be improved. We probably want to address this before merging, but I would like to focus on the API first.

hackermd replied:
> Why not all? Shouldn't returned results match all filters? That seems intuitive to me

Yeah, you are right. Good catch! Fixed via ed4383bd1dfec622008613f17676d93dbd92f7c1

hackermd replied:
I am leaving this comment unresolved as a reminder that we need more unit tests for these methods

hackermd commented on src/highdicom/sr/templates.py:
```diff
@@ -1935,20 +2888,515 @@ def __init__(
+            if len(matches) == 0:
+                sequences.append(seq)
+            else:
+                if any(matches):
```
Fixed via ed4383b

hackermd commented on src/highdicom/sr/templates.py:
```diff
@@ -1935,20 +2888,515 @@ def __init__(
+
+        return sequences
+
+    def get_image_measurments(
```
Good idea!

hackermd replied:
We may want to consider this for the other methods as well. However, for the planar/volumetric ROI methods, it would only work for the segmentation source images

CPBridge replied:
Yes, agree for planar ROIs. Not sure whether it makes sense for volumetric ROIs, which can straddle multiple frames? Could be convinced either way

hackermd replied:
What did we intend`reference_type` to be again? The name of the content item (e.g., `pydicom.sr.codes.DCM.ImageRegion`)?

CPBridge replied:
I think so. Or I think you were going to ponder having an enum whose values are codes

hackermd replied:
Take a look at 36a3f709e444b9eb1735de9b5a990b92ba943d95. What do you think?

hackermd replied:
Added unit tests for the planar ROI queries via c9fd35b5057fdcff6afd9c2599470ad1ec10c092

hackermd commented on src/highdicom/sr/templates.py:
```diff
@@ -1935,20 +2888,515 @@ def __init__(
+        graphic_type: Optional[
+            Union[GraphicTypeValues, GraphicTypeValues3D]
+        ] = None
+    ) -> List[PlanarROIMeasurementsAndQualitativeEvaluations]:
```
Agreed! See comment above :)

CPBridge commented:
> I have been having a hard time with this. The fact that the ROIs can be encoded "per value" as well as "per reference" complicates things quite a bit. I like the idea of allowing the caller to filter groups by ROI value type (SCOORD, SCOORD3D, IMAGE, COMPOSITE). 

I agree this is tricky. I don't think there is a perfect solution. I see a couple of problems with the proposed `value_type` based filtering:
- Ideally people should be able to use the library without having to understand how exactly the objects (`ImageRegion`, `ReferencedSegment`, etc) are encoded in DICOM under the hood. This suggestion would break that and require people to know which specific value type these types of regions are recorded using, which is not at all obvious. For example it is not at all clear without significant understanding of SRs that `COMPOSITE` would return ROI with a referenced real world value map. If we went forward with this, we would have to have a very good explanation in the docs I think.
- Filtering by value type does not allow you to distinguish between `ImageRegion3D` and `VolumeSurface`. This is probably not so serious, but potentially a bit confusing.

Two other options I see are:
- Introduce a new enum called `ROIType` with values `image_region`, `image_region_3d`, `volume_surface`, `segment`, `real_world_value_map`, and use that as a basis for filtering. This to me would be far more intuitive for users to understand. The disadvantage is that it creates an enum which does not really relate to anything in the standard.
- Have several different user-facing methods, something like:
  - `get_image_region_measurement_groups`
  - `get_segment_measurement_groups`
  - `get_volume_surface_measurement_groups`
  - ...
  They would probably all call the same method under the hood. This avoids the above problem and is possibly the most intuitive, but creates a lot of extra code. It may also allow other filtering options tuned specifically to each roi type within each of the methods (e.g. `get_segment_measurement_groups` could have a `segment_number` parameter).

Any thoughts?

Even trickier in my mind is the return type of the `.roi` property. What are you thinking there?

hackermd commented on tests/test_sr.py:
```diff
@@ -377,6 +391,8 @@ def test_datetime_item_construction_from_datetime(self):
         assert i.ConceptNameCodeSequence[0] == name
         assert i.DateTime == DT(value)
+        assert i.name == CodedConcept(*name)
+        assert i.value == value
```
This will need a bit of work. We currently use `unittest.TestCase`, which is incompatible with `pytest.fixture` (see [docs](https://docs.pytest.org/en/6.2.x/unittest.html#pytest-features-in-unittest-testcase-subclasses))

hackermd replied:
This particular class wouldn't need to be subclassed from `unittest.TestCase`, but we should then change this across the board.

CPBridge replied:
Hmm... I see. Well we already have several tests that do not use `unittest.TestCase`, for example those in `test_valuerep` and `test_utils`, so although it's a bit messy, it doesn't seem like a little more mix and match would hurt too much.

I've just been bitten too many times before by this `datetime_conversion` thing and really think having the test would be a good idea...

hackermd replied:
Ok. Maybe we should just rip out `unittest.TestCase`. This caused problems before.

CPBridge replied:
Perhaps, but I don't see any urgency.

hackermd commented on src/highdicom/sr/templates.py:
```diff
@@ -36,6 +44,306 @@
+    return False
+
+
+def _contains_volumetric_rois(group_item: Dataset) -> bool:
```
Addressed via 2ee628b065d8303c2dd1124bb8784f4e44f110f1

hackermd commented on src/highdicom/sr/templates.py:
```diff
@@ -584,6 +892,92 @@ def __init__(
+        return self[0].value
+
+    @property
+    def login_name(self) -> Union[str, None]:
```
I personally prefer `Union[str, None]` over `Optional[str]` for return values. I tend argue that there is no "optional" return value, since a function always returns a value. That value may be `None`.
In addition, the `Union[..., None]` style more naturally extends to more than two options (e.g., `Union[int, float, None]`). Unless you have a strong opinion about this, I would keep the current style.

CPBridge replied:
Ok, don't necessarily agree but we'll go with `Union` like we did in #74

hackermd commented on src/highdicom/sr/templates.py:
```diff
@@ -1366,6 +2131,133 @@ def __init__(
+        return None
+
+    @property
+    def measurements(self) -> List[NumContentItem]:
```
Hmm...I generally agree and like the idea. However, we would then also need to change this on `PlanarROIMeasurementsAndQualitativeEvaluations` and `VolumetricROIMeasurementsAndQualitativeEvaluations`. The problem that I see with that is that instances of these classes are returned by the `get_planar_roi_measurement_groups()` and `get_volumetric_roi_measurement_groups()`, respectively.

At the moment one can do

```python
for group in get_planar_roi_measurement_groups():
    print(group.measurements)
```

This would then have to become

```python
for group in get_planar_roi_measurement_groups():
    print(group.get_measurements())
```
which is not so nice IMO. What do you think?

CPBridge replied:
Well I agree the second one looks a bit nicer but it's pretty marginal and the question is how often are you really going to just be iterating through all measurements found in a group? I guess it depends on what you are doing but I would think in most of my use cases (ML focused of course) I would be looking for a particular measurement to do something with and would not be interested in whatever other measurements may be there at that point. And the difference here becomes:

```python
for meas in group.measurements:
    if meas.name == codes.SCT.Diameter:
        diameter = meas.value
        break
```

vs 

```python
diameter = group.get_measurements(codes.SCT.Diameter)[0].value
```
which to me trumps the minor difference above.

Of course, we could also decide to have both?

hackermd replied:
Or we expose the measurements as `Dict[pydicom.sr.coding.Code: List[highdicom.sr.NumContentItem]]`?

```python
diameter = group.measurements[codes.SCT.Diameter][0].value
```

CPBridge replied:
Hmm, that would have some advantages. But what if I wanted to filter measurements by finding site, tracking ID, or referenced image?

hackermd replied:
Implemented via 0b5ae3de258abb2c7c17ed3433fe607c9ece0d25 and e1c07cd0fb18ba66deb57106bd62f1b7f48865e8 and

hackermd commented on src/highdicom/sr/value_types.py:
```diff
@@ -64,23 +139,67 @@ def __setattr__(self, name: str, value: Any) -> None:
+            item.ContentSequence = ContentSequence.from_sequence(
+                dataset.ContentSequence
+            )
+        return item
```
Should be addressed via 4824abb477d2f5485b68743af82016fc5da29321

hackermd commented on src/highdicom/sr/value_types.py:
```diff
@@ -204,6 +362,34 @@ def __init__(
+
+        """
+        _assert_value_type(dataset, ValueTypeValues.CODE)
+        return super(CodeContentItem, cls)._from_dataset(dataset)
```
Addressed via 4ff586e5557d6f0ac2a3bbbb00834bb2818afddd

hackermd commented on src/highdicom/sr/value_types.py:
```diff
@@ -817,3 +1341,33 @@ def __init__(
+                return self.ReferencedTimeOffsets
+            except AttributeError:
+                return self.ReferencedDateTime
+
```
Added via 7c2981f36abfc8bf45ec21b173d5f333021da042

hackermd commented on src/highdicom/sr/templates.py:
```diff
@@ -36,6 +43,306 @@
+    )
+    regions_in_space_items = find_content_items(
+        group_item,
+        name=codes.DCM.RegionInSpace,
```
This is a recent addition to the standard and not yet in the latest pydicom version. Addressed via 2ee628b0

hackermd commented:
@CPBridge I addressed most of your comments and included your suggestions. There are a few remaining open comments that we would need to discuss.

hackermd commented on src/highdicom/sr/templates.py:
```diff
@@ -1366,6 +2131,133 @@ def __init__(
+        )
+
+    @property
+    def qualitative_evaluations(self) -> List[CodeContentItem]:
```
Maybe also `Dict[pydicom.sr.coding.Code, List[highdicom.sr.CodeContentItem]]`?

CPBridge replied:
Unlike for measurements, this should work quite nicely, however I also feel that the measurements and qualitative evaluations should have the same interface

hackermd replied:
Implemented via 0b5ae3de258abb2c7c17ed3433fe607c9ece0d25 and e1c07cd0fb18ba66deb57106bd62f1b7f48865e8.

CPBridge commented:
Merged to dev branch via command line to continue refinement there alongside #74
